modules/core/src/main/java/org/apache/ignite/internal/processors/query/GridQueryProcessor.java

```diff
@@ -974,7 +979,10 @@ public void onCacheStart0(GridCacheContextInfo<?, ?> cacheInfo, QuerySchema sche
                             if (!worker.nop()) {
                                 IgniteInternalFuture fut = worker.future();
 
-                                assert fut.isDone();
+                                // If the future is not complete it means that schema operation will be processed
+                                // by handle discovery message.
+                                if (!fut.isDone())
```

Comment:
why not fut.listen ? 
this part is called from **exchange-worker** and hold **stateMux** while other part and fut.onDone is called after **processSchemaOperationLocal** which also want this mutex, check : 
else if (op instanceof SchemaAddQueryEntityOperation) { ->
GridQueryProcessor#registerCache0

Reply:
Looks like `processSchemaOperationLocal` and `onCacheStart0` do the same steps in case start registered cache on join client.
What we should do with an error at the code is places in the fut.listen?